sidecar/src/controller.rs

```diff
@@ -0,0 +1,76 @@
+#![allow(dead_code)] // TODO remove when it is implemented
+
+use thiserror::Error;
+use tokio::select;
+use tokio::sync::oneshot::Receiver;
+use tokio::time::{Instant, Interval};
+
+use crate::config::Config;
+
+/// Sidecar operator controller
+#[derive(Debug)]
+pub(crate) struct Controller {
+    /// Sidecar operator config
+    config: Config,
+    /// Check interval
+    check_interval: Interval,
+}
+
+/// All possible errors
+#[derive(Error, Debug, PartialEq)]
+pub(crate) enum Error {
+    /// Graceful shutdown error
+    #[error("graceful shutdown")]
+    Shutdown,
+}
+
+/// Controller result
+type Result<T> = std::result::Result<T, Error>;
+
+impl Controller {
+    /// Constructor
+    pub(crate) fn new(config: Config) -> Self {
+        let check_interval = tokio::time::interval(config.check_interval);
+        Self {
+            config,
+            check_interval,
+        }
+    }
+
+    /// Perform a reconciliation
+    #[allow(clippy::integer_arithmetic)] // this error originates in the macro `tokio::select`
+    pub(crate) async fn reconcile_once(
+        &mut self,
+        shutdown_rx: &mut Receiver<()>,
```

Comment:
I recommend using event-listener as a shutdown trigger.

Reply:
I'll handle this in PR #11